### PR TITLE
docs(lua): more improvements

### PR DESCRIPTION
**Pull request by lewis6991**
### Can now generate documentation for lua tables

This works by looking for an assignment of the form `foo.bar = ...`.

We render these in doxygen with `table` return type and then `gen_vimdoc.py` now has logic to render these.

### No need for `@private` on local functions

Local functions are now rendered out with the `local_function` return type and `gen_vimdoc.py` will ignore these.

### Added `@nodoc` 

to disable documentation generation without marking them `@private`

### CODE REVIEW

File: runtime/lua/vim/loader.lua [hunk 1]
```diff
@@ -21,7 +21,10 @@ local M = {}
 
 ---@alias LoaderStats table<string, {total:number, time:number, [string]:number?}?>
 
+---@nodoc
 M.path = vim.fn.stdpath('cache') .. '/luac'
+
+---@nodoc
```
Comment (justinmk) [inline runtime/lua/vim/loader.lua:27]: Let's mention `@nodoc` at `:help dev-lua-doc`. Also `@meta` . 

We don't need a complete list there, but the custom or unusual ones are worth mentioning.

File: scripts/gen_vimdoc.py [hunk 2]
```diff
@@ -842,6 +842,15 @@ def extract_from_xml(filename, target, width, fmt_vimhelp):
...
             continue
 
+        if 'local_function' in return_type:
+            continue
+
+        if not CONFIG[target].get('include_tables', True):
+            if return_type.startswith('table'):
+                continue
+
+        istbl = return_type.startswith('table')
```
Comment (justinmk) [inline scripts/gen_vimdoc.py:852]:
```suggestion
        if 'local_function' in return_type:  # Special from lua2dox.lua.
            continue

        istbl = return_type.startswith('table')  # Special from lua2dox.lua.
        if istbl and not CONFIG[target].get('include_tables', True):
            continue
```

File: scripts/lua2dox.lua [hunk 3]
```diff
@@ -327,6 +328,10 @@ local function process_function_header(line)
       .. fn:sub(paren_start + 1)
   end
 
+  if line:match('local') then
+    return 'local_function ' .. fn .. '{}'
```
Comment (justinmk) [inline scripts/lua2dox.lua:332]:
```suggestion
    # Special: tell gen_vimdoc.py this is a local function.
    return 'local_function ' .. fn .. '{}'
```

File: scripts/lua2dox.lua [hunk 4]
```diff
@@ -348,6 +356,13 @@ local function process_line(line, in_stream, generics)
     return process_function_header(line)
   end
 
+  if not line:match('^local') then
+    local v = line_raw:match('^([A-Za-z][.a-zA-Z_]*)%s+%=')
+    if v and v:match('%.') then
+      return 'table '..v..'() {}'
```
Comment (justinmk) [inline scripts/lua2dox.lua:362]:
```suggestion
      # Special: gen_vimdoc.py has special handling for tables.
      return 'table '..v..'() {}'
```

**Review by justinmk — APPROVED** [review]

File: runtime/doc/lua.txt [hunk 5]
```diff
@@ -1419,11 +1353,81 @@ Option:remove({value})                                      *vim.opt:remove()*
...
+        vim.o.cmdheight = 4
+        print(vim.o.columns)
+        print(vim.o.foo)     -- error: invalid key
+<
+
+vim.wo                                                                *vim.wo*
+    Get or set window-scoped |options| for the window with handle {winid} and
+    buffer with number {bufnr}. Like `:setlocal` if {bufnr} is provided, like
+    `:set` otherwise. If [{winid}] is omitted then the current window is used.
+    Invalid {winid}, {bufnr} or key is an error.
```
Comment (zeertzjq) [inline runtime/doc/lua.txt:1413]: Now the `{winid}` and `{bufnr}` here appear out of nowhere.

File: runtime/doc/lua.txt [hunk 6]
```diff
@@ -1419,11 +1353,81 @@ Option:remove({value})                                      *vim.opt:remove()*
     Parameters: ~
       • {value}  (string) Value to remove
 
+vim.bo                                                                *vim.bo*
+    Get or set buffer-scoped |options| for the buffer with number {bufnr}.
+    Like `:set` and `:setlocal`. If [{bufnr}] is omitted then the current
+    buffer is used. Invalid {bufnr} or key is an error.
```
Comment (zeertzjq) [inline runtime/doc/lua.txt:1359]: Same for the `{bufnr}` here.

**Pull request merged**